src/document.cpp

```diff
@@ -1666,6 +1669,19 @@ rvoe<CapyPDF_TransparencyGroupId> PdfDocument::add_transparency_group(PdfDrawCon
     return CapyPDF_TransparencyGroupId{(int32_t)transparency_groups.size() - 1};
 }
 
+rvoe<CapyPDF_SoftMaskId> PdfDocument::add_soft_mask(CapyPDF_TransparencyGroupId tgid) {
+    auto id = add_object(FullPDFObject{std::format(R"(<<
+  /Type /Mask
+  /S /Luminosity
+  /G {} 0 R
+>>
```

Comment:
The basic outline of the change is good, but this hardcodes the mask type, which is not ok, we need to support all formats. This requires a new top level object. Thus the implementation needs to do roughly:

- add a new struct to `pdfcommon.hpp` called `SoftMask` that has entries S and G (and maybe comment to add BC and TR later)
- the constructor function should take both of those as arguments
- there needs to be a new enum for choosing between luminosity and alpha
- eventually the API should have functions for setting the optional entries, like `set_BC` and `set_TR`, but they are not needed for this, they can be added once someone actually needs them

Reply:
Thanks for the notes, this helped reformat the patch.